### PR TITLE
Documentation: Rework architecture guide (endpoints & multi node networking)

### PR DESCRIPTION
**Pull request by tgraf**
Also removes some outdated documentation around policy

### CODE REVIEW

File: Documentation/architecture.rst [hunk 1]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+
+Cilium makes application containers available on the network by assigning them
+IP addresses. Multiple application containers can share the same IP address, a
+typical example for this model is a Kubernetes Pod_. All application containers
+which share a common address are grouped together in what Cilium refers to as
+an *endpoint*.
+
+Allocating individual IP addresses enables the use of the entire Layer 4 port
+range by each *endpoint*. This essentially allows multiple application
+container running on the same cluster node to all bind to well known ports such
```
Comment (ianvernon) [inline Documentation/architecture.rst]: multiple application containers

File: Documentation/architecture.rst [hunk 2]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
 
-******************
-Address Management
-******************
+*********
+Endpoints
+*********
+
+Cilium makes application containers available on the network by assigning them
+IP addresses. Multiple application containers can share the same IP address, a
```
Comment (ianvernon) [inline Documentation/architecture.rst]: same IP address; a typical example...

File: Documentation/architecture.rst [hunk 3]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+an *endpoint*.
+
+Allocating individual IP addresses enables the use of the entire Layer 4 port
+range by each *endpoint*. This essentially allows multiple application
+container running on the same cluster node to all bind to well known ports such
+`80` without causing any conflicts.
+
+The default behaviour of Cilium is to assign both an IPv6 and IPv4 address to
+every *endpoint*. However, this behaviour can be configured to only allocate an
+IPv4 or IPv6 address. If both an IPv6 and IPv4 address are assigned, either
```
Comment (ianvernon) [inline Documentation/architecture.rst]: would be helpful to link to a section in the Docs that describes such configuration

File: Documentation/architecture.rst [hunk 4]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+address can be used to reach the endpoint. The same behaviour will apply with
+regard to policy rules, loadbalancing, etc.
+
+Identification
+==============
+
+For identification purposes, Cilium assign an *endpoint id* to all endpoints on
```
Comment (ianvernon) [inline Documentation/architecture.rst]: assigns

File: Documentation/architecture.rst [hunk 5]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
 
 ::
 
-    Cluster prefix: f00d::/64
+    $ docker run --net cilium -d -l app=benchmark tgraf/netperf
+    aaff7190f47d071325e7af06577f672beff64ccc91d2b53c42262635c063cf1c
+    $  cilium endpoint list
+    ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                   IPv4            STATUS
+               ENFORCEMENT
+    62006      Disabled      257        container:app=netperf         f00d::a00:20f:0:f236   10.15.116.202   ready
```
Comment (ianvernon) [inline Documentation/architecture.rst]: Shouldn't the label be "container:app=benchmark"?

File: Documentation/architecture.rst [hunk 6]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
-unspecified, Cilium will try and generate a unique prefix using the format
-``10.X.0.0/16`` where X is replaced with the last byte of the first global
-scope IPv4 address discovered on the node. This generated prefix is relatively
-weak in uniqueness so it is highly recommended to always specify the IPv4
-range.
+Cilium refers to a node as an individual member of a cluster. Each node must be
+running the ``cilium-agent`` and will operate in a mostly autonomous manner.
+Synchronization of state between Cilium agent's running on different nodes is
+kept to a minimum for simplicity and scale. It occurs exclusively via the
+Key-Value store or with packet metdata.
```
Comment (ianvernon) [inline Documentation/architecture.rst]: metadata

File: Documentation/architecture.rst [hunk 7]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+Address Management
+******************
+
+The address management is designed with simplicity and resilience in mind. This
+is achieved by delegating the address allocation for *endpoints* to each
+individual node in the cluster. Each cluster node is assigned a *node address
+allocation prefix* out of an overarching *cluster address prefix* and will
+allocate IPs for *endpoints* independently.
+
+This simplifies address handling and allows to make a fundamental assumption:
```
Comment (ianvernon) [inline Documentation/architecture.rst]: allows one to make

File: Documentation/architecture.rst [hunk 8]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
-transmit state and security context information between two container hosts by
-embedding the information in encapsulation headers or even unused bits of the
-IPv6 packet header. This allows Cilium to transmit the security context of
-where the packet origins from which allows tracing back which container labels
-are assigned to the origin container.
+Cilium is in full control over both ends of the connection for connections
+inside the cluster. It can thus transmit state and security context information
+between two container hosts by embedding the information in encapsulation
+headers or even unused bits of the IPv6 packet header. This allows Cilium to
+transmit the security context of where the packet origins from which allows
```
Comment (ianvernon) [inline Documentation/architecture.rst]: where the packet originates, which allows...

File: Documentation/architecture.rst [hunk 9]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
-  must be aware that each of the node IP prefixes are reachable by using the
-  node's primary IP address as an L3 next hop address.   In the case of a
-  traditional physical network this would typically involve announcing each
-  node prefix as a route using a routing protocol within the datacenter. Cloud
-  providers (e.g, AWS VPC, or GCE Routes) provide APIs to achieve the same result.
+- `Overlay / Encapsulation mode`_ 
+- `Direct / Native Routing Mode`_
 
 Regardless of the option chosen, the container itself has no awareness of the
 underlying network it runs on, it only contains a default route which points to
```
Comment (ianvernon) [inline Documentation/architecture.rst]: it runs on; it only...

File: Documentation/architecture.rst [hunk 10]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
-the IP address of the container host. Given the removal of the routing cache in
+the IP address of the cluster node Given the removal of the routing cache in
```
Comment (ianvernon) [inline Documentation/architecture.rst]: cluster node.

File: Documentation/architecture.rst [hunk 11]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
 the Linux kernel, this reduces the amount of state to keep to the per
```
Comment (ianvernon) [inline Documentation/architecture.rst]: amount of state to keep in the per-connection flow cache, which allows for the termination of millions of connections in each container.

File: Documentation/architecture.rst [hunk 12]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
 connection flow cache (TCP metrics) which allows to terminate millions of
 connections in each container.
 
+Overlay Network Mode
+====================
+
+When no configuration is provided, Cilium automatically runs in this mode.
+
+In this mode, all cluster nodes form a mesh of tunnels using the UDP based
+encapsulation protocols VXLAN_ or Geneve_. All container to container network
```
Comment (ianvernon) [inline Documentation/architecture.rst]: container-to-container

File: Documentation/architecture.rst [hunk 13]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+traffic is routed through these tunnels. This mode several major advantages:
```
Comment (ianvernon) [inline Documentation/architecture.rst]: ... mode has ...

File: Documentation/architecture.rst [hunk 14]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+
+- **Auto-configuration:** When running together with an orchestration system
+  such as Kubernetes, the list of all nodes in the cluster including their
+  associated allocation prefix node is made available to each agent
+  automatically. This means that if Kubernetes is being run with the
+  `--allocate-node-cidrs` option, Cilium can form an overlay network
+  automatically without any configuration by the user. New nodes joining the
+  cluster will automatically be incorporated into the mesh.
+
+- **Identity transfer:** Encapsulation protocols allow to carry arbitrary
```
Comment (ianvernon) [inline Documentation/architecture.rst]: allow for the carrying of arbitrary

File: Documentation/architecture.rst [hunk 15]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+Direct / Native Routing Mode
+============================
+
+.. note:: This is an advanced networking mode which requires the underlying
+          network to be made aware of container IPs. You can enable this mode
+          by running Cilium with the option ``--tunnel disabled``.
+
+In direct routing mode, Cilium will hand all packets which are not addressed
+for another local endpoint to the routing subsystem of the Linux kernel. This
+means that the packet will be routed just as if a local process would emit the
```
Comment (ianvernon) [inline Documentation/architecture.rst]: will be routed as if a local process had emitted the packet

File: Documentation/architecture.rst [hunk 16]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
+  prefix* via the node's IP to all other nodes.
+
+- Use of the cloud provider's routing functionality. Refer to the documentation
+  of your cloud provider for additional details. If you are running Kubernetes
+  with the `--cloud-provider` in combination with the `--allocate-node-cidrs`
+  option then this is configured automatically for IPv4 prefixes.
+
+.. note:: Use of direct routing mode currently only offers identity based
+          security policy enforcement for IPv6 where the security identity is
+          stored in the flowlabel. IPv4 is currently ot supported and thus
```
Comment (ianvernon) [inline Documentation/architecture.rst]: currently not supported

File: Documentation/architecture.rst [hunk 17]
```diff
@@ -394,13 +461,26 @@ are routable outside of the container cluster, communication with external
 hosts requires little more than enabling L3 forwarding on each of the Linux
 nodes.
 
-External Connectivity with Overlay Routing
-==========================================
+External Network Connectivity
+=============================
+
+If the destination of a packet relies outside of the cluster. Cilium will
```
Comment (ianvernon) [inline Documentation/architecture.rst]: lies outside of the cluster, Cilium will delegate...

File: Documentation/architecture.rst [hunk 18]
```diff
@@ -394,13 +461,26 @@ are routable outside of the container cluster, communication with external
...
+delegate routing to the routing subsystem of the cluster node to use the
+default route which is installed on the node of the cluster.
+
+As the IP addresses used for the **cluster prefix** are typically allocated
+from RFC1918 private address blocks and not publicly routeable. Cilium will
```
Comment (ianvernon) [inline Documentation/architecture.rst]: ...are not publicly routable

File: Documentation/architecture.rst [hunk 19]
```diff
@@ -394,13 +461,26 @@ are routable outside of the container cluster, communication with external
...
+the cluster. This behaviour can be disabled by running ``cililum-agent`` with
+the option ``--masquerade=false``.
 
-However, in the case of ''Overlay Routing'', accessing external hosts requires
-additional configuration.
+Public Endpoint Exposure
+========================
 
-In the case of containers accepting inbound connections, such services are
+In direct routing mode, endpoint IPs can be publicly routeable IPs and no
```
Comment (ianvernon) [inline Documentation/architecture.rst]: endpoint is referenced as \*endpoint\* elsewhere in the guide - should it be done so here as well for consistency?

File: Documentation/architecture.rst [hunk 20]
```diff
@@ -394,13 +461,26 @@ are routable outside of the container cluster, communication with external
...
+additional action needs to be taken.
+
+In overlay mode, endpoints accepting inbound connections, such services are
```
Comment (ianvernon) [inline Documentation/architecture.rst]: Same as above.

Comment (ianvernon) [inline Documentation/architecture.rst]: In overlay mode, endpoints that are accepting inbound connections are likely exposed via some kind of load-balancing layer...

**Review by ianvernon — CHANGES_REQUESTED** [review]
Great explanations! Just a few typos / grammatical mistakes that need to be fixed.

File: Documentation/architecture.rst [hunk 21]
```diff
@@ -208,174 +208,241 @@ label source defaults to ``any:`` which will match all labels regardless of
...
-  addressed to a local container and not addressed to the local node to the
-  Linux stack causing it to route the packet as it would route any other
-  non-local packet. As a result, the network connecting the Linux node hosts
-  must be aware that each of the node IP prefixes are reachable by using the
-  node's primary IP address as an L3 next hop address.   In the case of a
-  traditional physical network this would typically involve announcing each
-  node prefix as a route using a routing protocol within the datacenter. Cloud
-  providers (e.g, AWS VPC, or GCE Routes) provide APIs to achieve the same result.
+- `Overlay / Encapsulation mode`_ 
+- `Direct / Native Routing Mode`_
```
Comment (aalemayhu) [inline Documentation/architecture.rst]: The overlay link does not render properly.

**Review by aalemayhu — CHANGES_REQUESTED** [review]

File: Documentation/architecture.rst (hunk 5 above)
Comment (tgraf) [reply Documentation/architecture.rst]: Yes,fixed

**Pull request merged**